### PR TITLE
(feat) Release resolVI

### PR DESCRIPTION
**Pull request by canergen**

### CODE REVIEW

**Comment by canergen** [comment]
Failing tests due to genomepy.

File: src/scvi/external/resolvi/_model.py [hunk 1]
```diff
@@ -0,0 +1,651 @@
...
+        neighbor_key: str | None = None,
+        idx1: Sequence[int] | Sequence[bool] | None = None,
+        idx2: Sequence[int] | Sequence[bool] | None = None,
+        subset_idx: Sequence[int] | None = None,
+        mode: Literal["vanilla", "change"] = "change",
+        delta: float = 0.25,
+        batch_size: int | None = None,
+        fdr_target: float = 0.05,
+        silent: bool = False,
+        filter_outlier_cells: bool = False,
```
[the remark this answers is not in the document]
Comment (canergen) [reply src/scvi/external/resolvi/_model.py:515]: Fixed can you check our DE calls and add more arguments to de_dsp.

File: src/scvi/external/resolvi/_module.py [hunk 2]
```diff
@@ -0,0 +1,1142 @@
...
+    n_layers
+        Number of hidden layers used for encoder and decoder NNs
+    n_cats_per_cov
+        Number of categories for each extra categorical covariate
+    dispersion
+        One of the following
+
+        * ``'gene'`` - dispersion parameter of NB is constant per gene across cells
+        * ``'gene-batch'`` - dispersion can differ between different batches
+    log_variational
```
[the remark this answers is not in the document]
Comment (canergen) [reply src/scvi/external/resolvi/_module.py:60]: Removed it. See scvi for it. It's the default in all our models.

File: src/scvi/external/resolvi/_module.py [hunk 3]
```diff
@@ -0,0 +1,1142 @@
...
+        Number of spatial neighbors to consider for diffusion.
+    n_hidden
+        Number of nodes per hidden layer
+    n_latent
+        Dimensionality of the latent space
+    n_layers
+        Number of hidden layers used for encoder and decoder NNs
+    n_cats_per_cov
+        Number of categories for each extra categorical covariate
+    dropout_rate
```
[the remark this answers is not in the document]
Comment (canergen) [reply src/scvi/external/resolvi/_module.py:652]: dispersion is used throughout the function

[the remark this answers is not in the document]
Comment (canergen) [reply src/scvi/external/resolvi/_module.py:652]: Dropout rate removed. Sorry for the confusion.

File: src/scvi/model/base/_archesmixin.py [hunk 4]
```diff
@@ -145,15 +149,21 @@ def load_query_data(
...
             new_ten = new_state_dict[key]
+            load_ten = load_ten.to(new_ten.device)
             if new_ten.size() == load_ten.size():
                 continue
-            # new categoricals changed size
-            else:
-                dim_diff = new_ten.size()[-1] - load_ten.size()[-1]
-                fixed_ten = torch.cat([load_ten, new_ten[..., -dim_diff:]], dim=-1)
-                load_state_dict[key] = fixed_ten
+            fixed_ten = load_ten.clone()
```
[the remark this answers is not in the document]
Comment (canergen) [reply src/scvi/model/base/_archesmixin.py:155]: The old code is a special part. This is a more flexible Arches function (if people use additional n_batch dimensions.

File: src/scvi/model/base/_archesmixin.py [hunk 5]
```diff
@@ -170,6 +180,52 @@ def load_query_data(
 
         return model
 
+    @staticmethod
+    def _arches_pyro_setup(model, pyro_param_store):
```
[the remark this answers is not in the document]
Comment (canergen) [reply src/scvi/model/base/_archesmixin.py:184]: Essentially pyro manages it's own parameters. Loading torch parameters does not guarantee that the model behaves the same. One could write a model that only relies on torch but with major downsides. Loading a model requires loading the param_store.

File: src/scvi/model/base/_differential.py
[the remark this answers is not in the document]
Comment (canergen) [reply src/scvi/model/base/_differential.py:1]: Good with me.

File: src/scvi/model/base/_pyromixin.py
[the remark this answers is not in the document]
Comment (canergen) [reply src/scvi/model/base/_pyromixin.py:1]: I'm still not sure about Cell2location and other models that use scvi-tools. We should merge it and have a release candidate. Maybe ping Vitalii about it, who wrote most of the pyro code and is the author of Cell2location.

[the remark this answers is not in the document]
Comment (canergen) [reply src/scvi/model/base/_pyromixin.py:1]: Large parts of the code come from him though (but we're not merged and he didn't respond to requests to amend).

File: src/scvi/module/base/_base_module.py
[the remark this answers is not in the document]
Comment (canergen) [reply src/scvi/module/base/_base_module.py:1]: on_load now needs pyro_param_store. The models were never loading correctly.

File: src/scvi/external/resolvi/_model.py [hunk 6]
```diff
@@ -0,0 +1,651 @@
...
+        %(de_fdr_target)s
+        %(de_silent)s
+        **kwargs
+            Keyword args for :meth:`scvi.model.base.DifferentialComputation.get_bayes_factors`
+
+        Returns
+        -------
+        Differential expression DataFrame.
+        """
+        adata = self._validate_anndata(adata)
```
[the remark this answers is not in the document]
Comment (canergen) [reply src/scvi/external/resolvi/_model.py:461]: Not sure about it. Can you explain more about the benefits?

File: src/scvi/model/base/_archesmixin.py
[the remark this answers is not in the document]
Comment (canergen) [reply src/scvi/model/base/_archesmixin.py:1]: Can you help identify these locations? This would be unwanted side effects.

File: src/scvi/external/resolvi/_model.py [hunk 7]
```diff
@@ -0,0 +1,651 @@
...
+    from collections.abc import Iterable, Sequence
+    from typing import Literal
+
+    from anndata import AnnData
+
+logger = logging.getLogger(__name__)
+
+
+class RESOLVI(
+    PyroSviTrainMixin, PyroSampleMixin, PyroPredictiveMixin, BaseModelClass, ArchesMixin
```
[the remark this answers is not in the document]
Comment (canergen) [reply src/scvi/external/resolvi/_model.py:44]: It's very ResolVI specific function. We can call it ResolVIPredictiveMixin now.

File: docs/tutorials/notebooks
[the remark this answers is not in the document]
Comment (canergen) [reply docs/tutorials/notebooks:1]: Can you elaborate etc?

[the remark this answers is not in the document]
Comment (ori-kron-wis) [reply docs/tutorials/notebooks:1]: I added the required things for the tutorial to be linked correctly to RTD mainly. also references/ cites.

File: src/scvi/external/resolvi/_model.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (ori-kron-wis) [reply src/scvi/external/resolvi/_model.py:461]: its not a different structure, but maybe we will want to validate that we actually have what it takes to setup the object with this adata, namely having the spatial data and/or coordinates as part of the check, and if spatial data missing issue an error. 

on other hand, other spatial models don't have it (but thei inherit the base model class)
its just a suggestion.

File: src/scvi/external/resolvi/_model.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (ori-kron-wis) [reply src/scvi/external/resolvi/_model.py:515]: done

File: src/scvi/model/base/_pyromixin.py
[the remark this answers is not in the document]
Comment (ori-kron-wis) [reply src/scvi/model/base/_pyromixin.py:1]: will keep this in mind

File: src/scvi/model/base/_differential.py
[the remark this answers is not in the document]
Comment (ori-kron-wis) [reply src/scvi/model/base/_differential.py:1]: yes except that we added now more stuff to DE docstring.
Lets keep this in mind (in such case the other branches will be update from this addition)

File: src/scvi/model/base/_archesmixin.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (ori-kron-wis) [reply src/scvi/model/base/_archesmixin.py:184]: good. so a load/save test is essential.
